File: euler96.py
# sudoku solver
import sys
sys.stdin = open('sudoku.txt', 'r')
sudokus = list()
op = ''
for line in sys.stdin:
	if "Grid" in line:
		op+='*'
	else:
		op+=line
	
temp = op.split('*')
temp = temp[1:]

for each in temp:
	each = each.split('\n')
	#each = each[:-1]
	each = [[eval(x) for x in y] for y in each]
	sudokus.append(each)


def check(sudoku, xx, yy, t):
    a = 3*(xx/3)
    b = 3*(yy/3)
    #check in the box
    for x in range(a, a+3):
        for y in range(b, b+3):
            if x!=xx and y!=yy and sudoku[x][y]==t:
                return False
    #check in the row
    for x in range(9):
        if x!=xx and sudoku[x][yy]==t:
            return False
    #check in the column
    for y in range(9):
        if y!=yy and sudoku[xx][y]==t:
            return False
    return True

def noEmpty(sudoku):
    for x in range(9):
        for y in range(9):
            if sudoku[x][y]==0:
                return False
    return True

def solve(sudoku):
    if noEmpty(sudoku):
        return [True, sudoku]
    for x in range(9):
        for y in range(9):
            if sudoku[x][y]==0:
                for t in range(1,10):
                    sudoku[x][y] = t
                    if check(sudoku, x, y, t):
                        if solve(sudoku)[0]==True:
                            return [True, sudoku]
                sudoku[x][y]=0
                return [False, sudoku]

def printsudoku(sudoku):
	for x in sudoku:
		for y in x:
			print y,
		print '\n'
	
print len(sudokus[-1])
#assert False
# solve the sudokus here
s=0
cnt = 0
for x in sudokus:
	cnt+=1
	solve(x)
	printsudoku(x)
	print "***", cnt
	s+=(eval(str(x[0][0])+str(x[0][1])+str(x[0][2])))
print s

File: euler96.cpp
#include <iostream>
#include <fstream>

bool check(int* sudoku, int index);

bool solve(int* sudoku)
{
    for(int i=0;i<81;i++)
        if(sudoku[i]==0)
            goto notcomplete;
    return true;

    notcomplete:
       for(int i=0;i<81;i++)
       {
           if(sudoku[i]==0)
           {
               for(int j=1;j<10;j++)
               {
                   sudoku[i] = j;
                   // check for validity
                   if(check(sudoku,i))
                   {
                       if(solve(sudoku))
                           return true;
                   }
               }
               sudoku[i] = 0;
               return false;
           }
       }
}

bool check(int* sudoku, int index)
{
    int val = sudoku[index];

    // check in the row
    int start = (index/9)*9;
    for(int i=start;i<start+9;i++)
        if(i!=index and sudoku[i]==val)
            return false;

    // check in the column
    start = index%9;
    for(int i=start;i<81;i+=9)
        if(i!=index and sudoku[i]==val)
            return false;

    // check in the box
    int cnt =0;
    for(start=(index/27)*27;cnt<27;start++)// bounding
    {
        cnt++;
        if((start%9)/3 == (index%9)/3) // in the bound, so check
        {
            if(sudoku[start] == val and start!=index )
                return false;
        }
    }
    return true;
}

int main()
{
    std::ifstream f("sudoku.txt");
    std::string s;
    getline(f,s);// first line
    int count =0;
    int index=0;
    int sudoku[81];
    while(getline(f,s))
    {
        std::cout << s<<std::endl;;
        for(int k=0;k<s.length();k++)
        {
            sudoku[index]=s[k]-'0';
            index++;
        }
        count+=1;
        if(count>8)break;
    }
    for (int i=0;i<81;i++)
        std::cout << sudoku[i] << " ";
    std::cout << "\n";
    solve(sudoku);
    for (int i=0;i<81;i++)
    {
        std::cout << sudoku[i] << " ";
        if((i+1)%9==0)std::cout << "\n";
    }
    return 0;
}
